lib/spack/llnl/util/tty/log.py

```diff
@@ -89,180 +109,233 @@ def __enter__(self):
             termios.tcsetattr(fd, termios.TCSADRAIN, self.new_cfg)
 
         except Exception:
-            pass  # Some OS's do not support termios, so ignore.
+            pass  # some OS's do not support termios, so ignore
 
     def __exit__(self, exc_type, exception, traceback):
-        # If termios was avaialble, restore old settings after the
-        # with block
+        """If termios was avaialble, restore old settings."""
         if self.old_cfg:
             import termios
             termios.tcsetattr(
                 self.stream.fileno(), termios.TCSADRAIN, self.old_cfg)
 
 
 class log_output(object):
-    """Spawns a daemon that reads from a pipe and writes to a file
+    """Context manager that logs its output to a file.
+
+    In the simplest case, the usage looks like this::
+
+        with log_output('logfile.txt'):
+            # do things ... output will be logged
+
+    Any output from the with block will be redirected to ``logfile.txt``.
+    If you also want the output to be echoed to ``stdout``, use the
+    ``echo`` parameter::
+
+        with log_output('logfile.txt', echo=True):
+            # do things ... output will be logged and printed out
 
-    Usage::
+    And, if you just want to echo *some* stuff from the parent, use
+    ``force_echo``:
 
-        # Spawns the daemon
-        with log_output('logfile.txt', 'w') as log_redirection:
-           # do things ... output is not redirected
-           with log_redirection:
-                # do things ... output will be logged
+        with log_output('logfile.txt', echo=False) as logger:
+            # do things ... output will be logged
 
-    or::
+            with logger.force_echo():
+                # things here will be echoed *and* logged
```

Comment:
I almost added this and `with no_logging`, but I think we could easily add those when the need arises.  We might need to adjust the protocol a bit, and instead of using XON/XOFF to delineate echo'd regions of actual log, we'd use XON/OFF to start/stop some control protocol.  e.g.:

```shell
\xonTF\xoff\n    # echo, no log
\xonTT\xoff\n    # echo and log
\xonFT\xoff\n    # no echo, log
```
Personally, I'd save this for later, as I don't think it is needed right now.